Tempura/Navigation/NavigationWitness.swift

```diff
@@ -0,0 +1,182 @@
+        dispatch(Show(identifiersToShow, animated: animated, context: context)).void
+      },
+      hide: { identifierToHide, animated, context, atomic in
+        dispatch(Hide(identifierToHide, animated: animated, context: context, atomic: atomic)).void
+      }
+    )
+  }
+}
+
+// MARK: - Mock
+
+extension NavigationWitness {
+  /// The mocked NavigationWitness.
+  public static func mocked(
+    appendTo navigations: Wrapped<[NavigationRequest]> = .init(initialValue: []),
+    showHandlers: [RouteElementIdentifier: (Bool, Any?) -> Promise<Void>] = [:],
+    hideHandlers: [RouteElementIdentifier: (Bool, Any?, Bool) -> Promise<Void>] = [:]
+  ) -> Self {
+    return .init(
+      show: { identifiersToShow, animated, context in
+        navigations.value.append(contentsOf: identifiersToShow.map { .show($0) })
+        return all(identifiersToShow.compactMap { showHandlers[$0]?(animated, context) }).void
+      },
+      hide: { identifierToHide, animated, context, atomic in
+        navigations.value.append(.hide(identifierToHide))
+        return hideHandlers[identifierToHide]?(animated, context, atomic) ?? Promise(resolved: ())
+      }
+    )
+  }
+}
+
+// MARK: - Spy
+
+/// A navigation witness that expends the live one, adding the functionalities of the mocked one.
+extension NavigationWitness {
+  public static func spy(
+    dispatch: @escaping AnyDispatch,
+    appendTo navigations: Wrapped<[NavigationRequest]> = .init(initialValue: []),
+    showHandlers: [RouteElementIdentifier: (Bool, Any?) -> Promise<Void>] = [:],
+    hideHandlers: [RouteElementIdentifier: (Bool, Any?, Bool) -> Promise<Void>] = [:]
+  ) -> Self {
+    let live: NavigationWitness = .live(dispatch: dispatch)
+    let mocked: NavigationWitness = .mocked(appendTo: navigations, showHandlers: showHandlers, hideHandlers: hideHandlers)
+
+    return .unimplemented(
+      show: { identifiersToShow, animated, context in
+        return live.show(identifiersToShow, animated: animated, context: context)
+          .then {
+            mocked.show(identifiersToShow, animated: animated, context: context)
+          }
+      },
+      hide: { identifierToHide, animated, context, atomic in
+        return live.hide(identifierToHide, animated: animated, context: context, atomic: atomic)
+          .then {
+            mocked.hide(identifierToHide, animated: animated, context: context, atomic: atomic)
+          }
+      }
+    )
+  }
+}
+
+// MARK: - Unimplemented
+
+extension NavigationWitness {
+  /// The unimplemented NavigationWitness.
+  public static func unimplemented(
+    show: @escaping ([RouteElementIdentifier], Bool, Any?) -> Promise<Void> = { _, _, _ in fatalError() },
+    hide: @escaping (RouteElementIdentifier, Bool, Any?, Bool) -> Promise<Void> = { _, _, _, _ in fatalError() }
+  ) -> Self {
+    return .init(
+      show: show,
+      hide: hide
+    )
+  }
+}
+
+// MARK: - Wrapped
+
+/// Wraps any object into a reference-type object
+public class Wrapped<T> {
```

Comment:
[i] If you look in Ramen there is a `Box` type that does this but it also has dynamic dispatch to the underlying value (which is kinda cool to have). I'd put that here

Reply:
Replaced `Wrapped` with ad hoc `NavigationRequests` type for improved ergonomic